### PR TITLE
Bump Go to 1.9.2

### LINKED ISSUE #35152: docker pull failing with failed to parse certificate: x509: unhandled critical extension (based on #31949)
Dear all,

I'm opening this one but it definitely follows #31949.
I add here informations and diagnostics we made while facing the problem.

**Description**

We get a "failed to parse certificate: x509: unhandled critical extension" error message when trying to pull images on public regitries.

The environment is RHEL 7 with all docker releases we tested (EE and CE) and in particular with: docker-ce-17.09.0. This is due to Name Constraints on the certificate generated by the local PKI (see further details below)

**Steps to reproduce the issue:**

1. Configure HTTPS_PROXY and HTTP_PROXY correctly in the system and in the systemd unit file for docker
2. ```docker pull centos```

**Describe the results you received:**

```
... failed to parse certificate: x509: unhandled critical extension
```
And no image pulled at all ;)

**Describe the results you expected:**

We should have pulled the image without error. (curl works fine in the same situation).

**Additional information you deem important (e.g. issue happens only occasionally):**

We are in a situation where security team put in place SSL interception with a local PKI. Each https request is intercepted and certificates are generated on the fly for the sites we browse. The local PKI as a CA and a sub CA. The TLS certificate of the sub CA which appears in the TLS exchange includes  "Name Constraints" as critical with something like this:

```
X509v3 Name Constraints: critical
                Excluded:
                  DNS:.localdom
```

After research, it seems that docker for RHEL7 is built with golang 1.8 and that this version of golang does not support Name Constraints of type Excluded. 

You can reproduce the exact same behavior by running the simple [example of upstream golang](https://golang.org/pkg/crypto/x509/#example_Certificate_Verify) and using any generated certificate with Name Constraints critical+excluded.

For example, this one to put in the certPEM in the golang example:

```
-----BEGIN CERTIFICATE-----
MIIDejCCAmKgAwIBAgIJAIc9PgrS4S4JMA0GCSqGSIb3DQEBCwUAMEIxCzAJBgNV
BAYTAkZSMQ0wCwYDVQQIDARQQUNBMRIwEAYDVQQHDAlNYXJzZWlsbGUxEDAOBgNV
BAoMB1Jvb3QgNDIwHhcNMTcxMDEwMTExMDQ4WhcNMjcxMDA4MTExMDQ4WjBCMQsw
CQYDVQQGEwJGUjENMAsGA1UECAwEUEFDQTESMBAGA1UEBwwJTWFyc2VpbGxlMRAw
DgYDVQQKDAdSb290IDQyMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA
7N9WtdUiLvN3a2InayZe2HGQKQCKVGsOM3NKMO2GnoWmZN5iA0OSNa4VDW8Lx5pH
nN+CIo87wD+boA1QyfkjJfkKzU6Dg1yoE5jFEPs3UJbkn/JQEmFlBI6rMio7Y7FW
feMXgQiKenqX44fDDgdLaEwFvg+B98mj2Md/sXywHw+k6cSIEZkRMK+/P+PTU0BY
B8P4wDsWmgpukN61KGnZg5FT7G+0rxAedf8Ix4JpGn/bE+86YL4sS5raI5yCjVom
TMLz2ZN6DzyAAmfxszQHz2faxA0Xlgcxd8cS0TB8wgAh4KrSe7rnZGGcnJt2hGLr
S+zIo25TJVKSFpe6qkPzdwIDAQABo3MwcTAdBgNVHQ4EFgQUsdPw4t8nn3dKWMbV
qrgepSKO9f4wHwYDVR0jBBgwFoAUsdPw4t8nn3dKWMbVqrgepSKO9f4wDwYDVR0T
AQH/BAUwAwEB/zAeBgNVHR4BAf8EFDASoRAwDoIMLmxvY2FsZG9tYWluMA0GCSqG
SIb3DQEBCwUAA4IBAQAsD1uiJCGb35NcuZupAo/5+LDFbt9R39HhMogHk1d2x1YW
0p+04M8MApS6Ki6j98IjgKsAZrRxO4k38Yj6cMUHMnE3DGE7myYimtYsB5lzhLPC
ubTfp83bJQvd66t8GArgyvrnEiSYV7zQVKchxGM25a6nrU9xn4DloP5fVYElanVD
xTevqzYpfRxglUHDQwyHalPHQinGQWelPZzQrzH7HXccyPl6a/bGTtwXSzrb8Z8N
UI1oqcW3u2VZuc65+6y41RLzEyD++lVudkSU2OVNluatQeMSj6/MYPhffD7Kj6Hr
JuSSB5Tg47pwyBdYZ1cOGZ9uq5eZfO93QIfDVoVM
-----END CERTIFICATE-----
```

This has been fixed in golang 1.9 [commit here](https://go-review.googlesource.com/c/go/+/36900)

**Output of `docker version`:**

```
Client:
 Version:      17.09.0-ce
 API version:  1.32
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:41:23 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.09.0-ce
 API version:  1.32 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:42:49 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 2
Server Version: 17.09.0-ce
Storage Driver: vfs
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 06b9cb35161009dcb7123345749fef02f7cea8e0
runc version: 3f2f8b84a77f73d38244dd690525642a72156c64
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.10.15-1-pve
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 1GiB
Name: docker-test-centos
ID: TKJ7:JIPX:WZAX:2DQG:7X5R:FKU2:XIPO:JPWS:37KK:NTV6:FGAJ:3LWL
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

It's a golang issue and so we'd like to know if you plan to build docker on RHEL7 with golang 1.9 which solves the problem or if you could ask RH to backport commit d1211b9 to golang 1.8 as included in RH so that you rebuild docker with this particular version.
Until that we can't see any other option than bypassing SSL interception which is not always possible in companies with security in mind.

Best and thank you for the good work!

### PR DESCRIPTION
**Pull request by thaJeztah**
let's see what fails (or not) 👍 

fixes https://github.com/moby/moby/issues/35152

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
interesting
```
08:48:07 --- FAIL: TestHashFile (0.07s)
08:48:07 	tarsum_test.go:66: invalid checksum. expected 1149ab94af7be6cc1da1335e398f24ee1cf4926b720044d229969dfc248ae7ec, got 55dfeb344351ab27f59aa60ebb0ed12025a2f2f4677bf77d26ea7a671274a9ca
08:48:07 --- FAIL: TestHashSubdir (0.05s)
08:48:07 	tarsum_test.go:103: invalid checksum. expected d7f8d6353dee4816f9134f4156bf6a9d470fdadfb5d89213721f7e86744a4e69, got 74a3326b8e766ce63a8e5232f22e9dd895be647fb3ca7d337e5e0a9b3da8ef28
08:48:07 FAIL
08:48:07 coverage: 15.8% of statements
08:48:07 FAIL	github.com/docker/docker/builder/remotecontext	0.333s
```

**Comment by thaJeztah** [comment]
ping @unclejack @tonistiigi ^^

**Comment by AkihiroSuda** [comment]
git bisect says golang/go@66b5a2f is the first commit causing this failure:

```
commit 66b5a2f3f0b1d94f69763114a85a977f5bc0277a
Author: Lars Jeppesen <jeppesen.lars@gmail.com>                
Date:   Sat Apr 29 23:25:34 2017 +0200  
                                                                                            
    archive/tar: remove file type bits from mode field                         
                           
    When writing tar files by using the FileInfoHeader                                   
    the type bits was set in the mode field of the header
    This is not correct according to the standard (GNU/Posix) and 
    other implementations.       
        
    Fixed #20150
                                                
    Change-Id: I3be7d946a1923ad5827cf45c696546a5e287ebba
    Reviewed-on: https://go-review.googlesource.com/42093
    Reviewed-by: Joe Tsai <thebrokentoaster@gmail.com>
    Run-TryBot: Joe Tsai <thebrokentoaster@gmail.com>
    TryBot-Result: Gobot Gobot <gobot@golang.org>
```

**Comment by tonistiigi** [comment]
@AkihiroSuda Thanks for the debug.

I'm not sure what's the best approach here. Do we want to break build cache for everyone or fork the `archive.FileInfoHeader` and keep using the wrong headers? Also related: moby/buildkit#38

**Comment by AkihiroSuda** [comment]
Although build cache are not persistent data, people will get surprised if caches were broken by just updating Go compiler.
So my opinion is to keep using the wrong header until we integrate buildkit to the engine, but I don't have any strong opinion.
WDYT?

**Comment by AkihiroSuda** [comment]
I'll try to open a separate PR for filling file bits

**Comment by AkihiroSuda** [comment]
opened https://github.com/moby/moby/pull/33935

**Comment by tiborvass** [comment]
Can we confirm that the impact is only on caching? What about tar layers?

**Comment by tonistiigi** [comment]
@tiborvass Layers use tar-split copy of the original headers and don't regenerate them.

**Comment by thaJeztah** [comment]
Rebased, because https://github.com/moby/moby/pull/33935 was merged; let's see if it goes green this time

**Comment by tophj-ibm** [comment]
z failure looks serious, can you take a look @michael-holzheu ?

from the log
```
SIGILL: illegal instruction
PC=0x283186 m=0 sigcode=2
```

**Comment by utzb** [comment]
The z golang issue is understood and addressed. Thanks for attempting in the first place!

**Comment by thaJeztah** [comment]
Thanks @utzb - is that fix upstream (golang), or an issue fixed in this repository? (IOW, should a rebase fix it, or do we need to wait for the next Go beta/release)

**Comment by utzb** [comment]
The fix will be in upstream golang. With a little luck, it will be in 1.9beta3, but for sure in the first rc.
I verified test-unit and test-integration-cli run smooth using a golang with an early version of the fix.

**Comment by thaJeztah** [comment]
Thanks, good to hear!

**Comment by utzb** [comment]
Just FYI, this is now https://github.com/golang/go/issues/21048 (https://go-review.googlesource.com/c/49250/)

**Comment by thaJeztah** [comment]
bumped to 1.9rc1

**Comment by tiborvass** [comment]
Since we are bumping Go versions, we should probably update archive/tar as well.
Here is the changelog for src/archive/tar: https://gist.github.com/tiborvass/f9303e496644c77bd11ba08ec00d9fa0

cc @tonistiigi @dmcgowan

**Comment by thaJeztah** [comment]
Cool, it's greeeeen!

**Comment by thaJeztah** [comment]
triggered ARM, as it didn't run yet

**Comment by tophj-ibm** [comment]
3 Arm failures:

1. DockerSuite.TestLogsFollowSlowStdoutConsumer which has been failing on ARM for as long as I can remember

2. DockerSwarmSuite.TestSwarmPublishDuplicatePorts. which was an old flaky test I haven't seen failed in a while #30427

3. What looks like an insecure registry problem with DockerPluginSuite.SetUpSuite, which I'm not why would only fail on arm if it's an issue. ping @cpuguy83 
```time="2017-07-27T04:03:13Z" level=error msg="Not continuing with push after error: Get https://127.0.0.1:5000/v2/: http: server gave HTTP response to HTTPS client" ```

**Comment by AkihiroSuda** [comment]
rc2 available now

**Comment by thaJeztah** [comment]
Bumped to 1.9rc2 (thanks for the ping @AkihiroSuda)

**Comment by thaJeztah** [comment]
@cpuguy83 this one is still failing; any ideas?

```
16:42:25 time="2017-08-15T16:42:25Z" level=error msg="Not continuing with push after error: Get https://127.0.0.1:5000/v2/: http: server gave HTTP response to HTTPS client" 
16:42:25 
16:42:25 ----------------------------------------------------------------------
16:42:25 FAIL: check_test.go:461: DockerPluginSuite.SetUpSuite
16:42:25 
16:42:25 check_test.go:469:
16:42:25     c.Assert(err, checker.IsNil, check.Commentf("failed to create plugin"))
16:42:25 ... value *errors.withStack = error pushing plugin: Get https://127.0.0.1:5000/v2/: http: server gave HTTP response to HTTPS client ("error pushing plugin: Get https://127.0.0.1:5000/v2/: http: server gave HTTP response to HTTPS client")
16:42:25 ... failed to create plugin
16:42:25 
```

**Comment by yongtang** [comment]
There are several warnings with respect to linkage:
```
14:45:15 # github.com/docker/docker/cmd/dockerd
14:45:15 /tmp/go-link-421380527/000010.o: In function `mygetgrouplist':
14:45:15 /usr/local/go/src/os/user/getgrouplist_unix.go:15: warning: Using 'getgrouplist' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
14:45:15 /tmp/go-link-421380527/000010.o: In function `mygetgrgid_r':
14:45:15 /usr/local/go/src/os/user/cgo_lookup_unix.go:38: warning: Using 'getgrgid_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
14:45:15 /tmp/go-link-421380527/000010.o: In function `mygetgrnam_r':
14:45:15 /usr/local/go/src/os/user/cgo_lookup_unix.go:43: warning: Using 'getgrnam_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
14:45:15 /tmp/go-link-421380527/000010.o: In function `mygetpwnam_r':
14:45:15 /usr/local/go/src/os/user/cgo_lookup_unix.go:33: warning: Using 'getpwnam_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
14:45:15 /tmp/go-link-421380527/000010.o: In function `mygetpwuid_r':
14:45:15 /usr/local/go/src/os/user/cgo_lookup_unix.go:28: warning: Using 'getpwuid_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
14:45:15 Created binary: bundles/17.06.0-dev/binary-daemon/dockerd-17.06.0-dev
14:45:15 Copying nested executables into bundles/17.06.0-dev/binary-daemon
```

The above warnings point to the usage of os/user in vendor/golang.org/x/oauth2/google/sdk.go.

This is loosely related to #29478 which avoids calling os/user, so the warnings may be OK (we may still want to remove the warnings though.)

**Comment by yongtang** [comment]
The following warnings are related to static build of runc:
```
14:43:11 CGO_ENABLED=1 go build -i -tags "seccomp apparmor selinux cgo static_build" -ldflags "-w -extldflags -static -X main.gitCommit="2d41c047c83e09a6d61d464906feb2a2f3c52aa4" -X main.version=1.0.0-rc3" -o runc .
14:43:18 # github.com/opencontainers/runc
14:43:18 /tmp/go-link-229926267/000008.o: In function `_cgo_a5acef59ed3f_Cfunc_dlopen':
14:43:18 /tmp/go-build/github.com/opencontainers/runc/vendor/github.com/coreos/pkg/dlopen/_obj/cgo-gcc-prolog:76: warning: Using 'dlopen' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
14:43:18 /tmp/go-link-229926267/000001.o: In function `_cgo_b0c710f30cfd_C2func_getaddrinfo':
14:43:18 /tmp/go-build/net/_obj/cgo-gcc-prolog:46: warning: Using 'getaddrinfo' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
14:43:18 CGO_ENABLED=1 go build -i -tags "seccomp apparmor selinux cgo static_build" -ldflags "-w -extldflags -static -X main.gitCommit="2d41c047c83e09a6d61d464906feb2a2f3c52aa4" -X main.version=1.0.0-rc3" -o contrib/cmd/recvtty/recvtty ./contrib/cmd/recvtty
```

Warning `getaddrinfo` could be addressed with `-tag "netgo"`.

Warning `dlopen` might be addressed by removing `cgo` tag for static build only (will need to check further to see if there are any implications.

**Comment by yongtang** [comment]
Warning `'getaddrinfo'` also appears in containerd build, think it might be addressed with `netgo` as well.

**Comment by yongtang** [comment]
By the way, 1.9 has been officially released.

**Comment by yongtang** [comment]
Created a PR opencontainers/runc#1577  in `runc` to address the 'getaddrinfo' warning.

**Comment by yongtang** [comment]
Created a PR opencontainers/runc#1579  in runc to address the dlopen warning.

**Comment by AkihiroSuda** [comment]
https://github.com/moby/moby/pull/33935 (tar compatibility issue) seems to need more discussion

**Comment by thaJeztah** [comment]
> By the way, 1.9 has been officially released.

Rebased, and bumped to Go 1.9

**Comment by yongtang** [comment]
Changed the name of the PR from `1.9rc2` -> `1.9`

**Comment by thaJeztah** [comment]
Thanks! forgot it this time :blush:

**Comment by thaJeztah** [comment]
Windows is failing on the same test again, so may be a genuine failure:

```
15:53:48 ----------------------------------------------------------------------
15:53:48 FAIL: docker_cli_cp_to_container_test.go:26: DockerSuite.TestCpToErrSrcNotExists
15:53:48 
15:53:48 running `docker cp  d:\CI\CI-e6e6312a4\test-cp-to-err-src-not-exists565326422\file1 c30610d2bda9a903a7706a67ade84cf08881c286ed24ccb20227e1ab71df250f:file1`
15:53:48 docker_cli_cp_to_container_test.go:38:
15:53:48     c.Assert(strings.ToLower(err.Error()), checker.Contains, strings.ToLower(srcStatErr.Error()))
15:53:48 ... obtained string = "error executing `docker cp` command: exit status 1: getfileattributesex d:\\ci\\ci-e6e6312a4\\test-cp-to-err-src-not-exists565326422\\file1: the system cannot find the file specified.\n"
15:53:48 ... substring string = "createfile d:\\ci\\ci-e6e6312a4\\test-cp-to-err-src-not-exists565326422\\file1: the system cannot find the file specified."
15:53:48 
```

@johnstep @jhowardmsft any ideas? ^^

**Comment by jhowardmsft** [comment]
@thaJeztah Fascinating. No, no idea at all. I'll start it again, but to my knowledge this isn't a flaky test. @darrenstahlmsft You ever seen this one fail?

**Comment by darrenstahlmsft** [comment]
That's a new failure to me. I know there were some somewhat related changes in Go1.9 around symlinks and opening files, possibly related to that.

**Comment by unclejack** [comment]
It'd be a good idea to wait until Go 1.9.1 is out before merging this PR. Go 1.9 has a bunch of bugs which could lead to some interesting and very hard to debug issues.

**Comment by thaJeztah** [comment]
Thanks for the heads-up @unclejack!

In the meantime we should look into that failure, and see what makes it fail

/cc @johnstep as well in case he has some spare cycles (doubtful) 😃

**Comment by kolyshkin** [comment]
As go1.9 now have a native arm64 (aka aarch64) binary build, it makes sense to use it rather than compiling from source. Please see  https://github.com/moby/moby/pull/34923/commits/6bd08f7a80b75bb110249b48c2af648e2d9fb2b2#diff-0a5836d551d7f223bdc8b46372ca0089

**Comment by kolyshkin** [comment]
> #33935 (tar compatibility issue) seems to need more discussion

Unfortunately `archive/tar` was not yet fixed in go 1.9, but probably will be in 1.9.1, as the [patch](https://github.com/golang/go/commit/01e45c736882754ca785b5a802ec0866a6544f8b) fixing https://github.com/golang/go/issues/21005 is already committed.

**Comment by thaJeztah** [comment]
PowerPC failing; looks like issues with the debian repositories;

```
18:43:43 W: Unable to read /etc/apt/apt.conf.d/ - DirectoryExists (2: No such file or directory)
18:43:43 W: GPG error: http://cdn-fastly.deb.debian.org/debian stretch-updates InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 8B48AD6246925553 NO_PUBKEY 7638D0442B90D010
18:43:43 W: The repository 'http://cdn-fastly.deb.debian.org/debian stretch-updates InRelease' is not signed.
18:43:43 W: GPG error: http://cdn-fastly.deb.debian.org/debian stretch Release: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 8B48AD6246925553 NO_PUBKEY 7638D0442B90D010 NO_PUBKEY EF0F382A1A7B6500
18:43:43 W: The repository 'http://cdn-fastly.deb.debian.org/debian stretch Release' is not signed.
18:43:43 W: GPG error: http://security.debian.org stretch/updates InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 9D6D8F6BC857C906 NO_PUBKEY 8B48AD6246925553
18:43:43 W: The repository 'http://security.debian.org stretch/updates InRelease' is not signed.
```

**Comment by tophj-ibm** [comment]
@thaJeztah something is up with power node6, I've been seeing it fail on other tests for the same reason. I temporarily removed it from the pool until I can figure out why it's throwing that.

**Comment by thaJeztah** [comment]
Thanks @tophj-ibm

**Comment by thaJeztah** [comment]
https://github.com/moby/moby/pull/33892#issuecomment-329977296 is still failing on Windows (probably expected)

**Comment by johnstep** [comment]
The problem for Windows is the `os.Stat` error changed (using file `C:\not-exist` below):

**1.8.3**
`GetFileAttributesEx C:\not-exist: The system cannot find the file specified.`

**1.9**
`CreateFile C:\not-exist: The system cannot find the file specified.`

However, the error from `docker cp` is still from `GetFileAttributesEx` when built with 1.9:
```
$ .\docker.exe cp C:\not-exist "(docker ps -aqn1):not-exist"
GetFileAttributesEx C:\\not-exist: The system cannot find the file specified.
```

Therefore, the error string does not contain the full sub-string from above, since the first function failing in `os.Stat` is now `CreateFile` instead of `GetFileAttributesEx`. We need to change how the test compares the error; the test is otherwise working as expected.

File: integration-cli/docker_cli_cp_to_container_test.go
Comment (thaJeztah) [inline integration-cli/docker_cli_cp_to_container_test.go]: Lets see if this works instead @johnstep  - problem was that `os.Stat()` returns a different kind of error than `runDockerCp()`, so not sure if this is correct 😅

**Comment by tophj-ibm** [comment]
@thaJeztah should we bump + test with go1.9.2? I can't recall any other issues this was blocked on.

**Comment by thaJeztah** [comment]
Yes, let me rebase this and bump (sorry, this dropped from my radar again 👍 )

**Comment by thaJeztah** [comment]
Updated 👍

File: contrib/builder/rpm/amd64/amazonlinux-latest/Dockerfile
Comment (thaJeztah) [inline contrib/builder/rpm/amd64/amazonlinux-latest/Dockerfile]: the generate script was updated in https://github.com/moby/moby/commit/b877fc31c51baaf1ec42501ab366dd05a5b6bb21, but looks like the script wasn't run to update the dockerfiles

**Comment by darrenstahlmsft** [comment]
I'm investigating a blocker on Windows that prevents pulling some images. Not sure how CI passed tbh.

**Comment by thaJeztah** [comment]
Removing the `[do not merge]` I think we can review this now; the changes made for the behaviour-change in Windows are a bit quirky, and looking at those, perhaps it should be a docker/cli test, so I added a TODO for that https://github.com/moby/moby/pull/33892/files#diff-f77ac0f89cdcfa4259ac67bced3d418cR39

@darrenstahlmsft was there still something to look into from your side?

ping @AkihiroSuda @vdemeester @cpuguy83 @yongtang PTAL

**Review by vdemeester — APPROVED** [review]
LGTM 🐮

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Comment by cpuguy83** [comment]
👎 Checking error strings, but the current error handling for this endpoint is not good enough as a true "not found" error would be seen as the container not being found, so we need more data in the error message to differentiate a "file not found" from a "container not found".

**Review by AkihiroSuda — APPROVED** [review]

**Review by yongtang — APPROVED** [review]

**Comment by thaJeztah** [comment]
@cpuguy83 agree it's pretty dirty; I didn't see a better option directly, other than splitting the test into a unit test (docker/cli), and an API test here (and may require changes to the endpoint itself as you mentioned).

Do you want me to look into this in this PR, or should we open an issue for tracking that effort?

**Comment by cpuguy83** [comment]
We don't need it for this PR.

**Review by dnephin — APPROVED** [review]
LGTM

**Review by unclejack — APPROVED** [review]
LGTM

**Comment by darrenstahlmsft** [comment]
@thaJeztah yes, Windows needs https://github.com/Microsoft/hcsshim/pull/144

**Comment by darrenstahlmsft** [comment]
Please hold the merge until that is in

**Comment by thaJeztah** [comment]
Thanks for checking @darrenstahlmsft - temporarily changed back to `WIP`

**Comment by thaJeztah** [comment]
https://github.com/moby/moby/pull/35554 was merged; rebased this PR, and removed "WIP" again; this should be ready to go now

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Review by vieux — APPROVED** [review]
LGTM

**Pull request merged**
